### PR TITLE
[Protocol3] allow withdrawal of balances not owned by users

### PR DESCRIPTION
**Pull request by dong77**
This is to address issue #201

### CODE REVIEW

File: packages/loopring_v3/contracts/impl/libexchange/ExchangeAdmins.sol [hunk 1]
```diff
@@ -256,4 +256,36 @@ library ExchangeAdmins
...
+            ERC20(token).balanceOf(address(this));
+
+        uint userBalance = S.tokenBalances[token];
+        assert(totalBalance >= userBalance);
+
+        amount = totalBalance - userBalance;
+
+        if (amount > 0) {
+            bool success;
+            if (token == address(0)) {
```
Comment (Brechtpd) [inline packages/loopring_v3/contracts/impl/libexchange/ExchangeAdmins.sol]: ETH case can be removed

Comment (dong77) [reply packages/loopring_v3/contracts/impl/libexchange/ExchangeAdmins.sol]: Why?

Comment (Brechtpd) [reply packages/loopring_v3/contracts/impl/libexchange/ExchangeAdmins.sol]: Oh sorry, I was confused, you're right.

**Comment by dong77** [comment]
@Brechtpd since you are most familiar with withdrawal/deposit logics, please make sure I updated `tokenBalances` correctly in the right places.

**Comment by Brechtpd** [comment]
> @Brechtpd since you are most familiar with withdrawal/deposit logics, please make sure I updated `tokenBalances` correctly in the right places.

I think deposits/withdrawals are correct, but there's a problem with ETH because it's used as the fee for on-chain requests (and can be withdrawn as the block fee for the operator), so ETH needs some extra logic.

**Comment by Brechtpd** [comment]
> I think deposits/withdrawals are correct, but there's a problem with ETH because it's used as the fee for on-chain requests (and can be withdrawn as the block fee for the operator), so ETH needs some extra logic.

Now that ETH cannot be deposited to the contract any more using the default function, and all other functions  which use ETH can never pay too much ETH (the surplus is immediately sent back) there is no way ETH stored in the contract is not owned by anyone (either by a user or for the block fee). The ETH case in `withdrawBalanceNotOwnedByUsers` should be removed and it should not be possible to withdraw ETH this way.

For tokens everything should work correctly as it is now. But let me think about this some more. :)

**Comment by dong77** [comment]
> > I think deposits/withdrawals are correct, but there's a problem with ETH because it's used as the fee for on-chain requests (and can be withdrawn as the block fee for the operator), so ETH needs some extra logic.
> 
> Now that ETH cannot be deposited to the contract any more using the default function, and all other functions which use ETH can never pay too much ETH (the surplus is immediately sent back) there is no way ETH stored in the contract is not owned by anyone (either by a user or for the block fee). The ETH case in `withdrawBalanceNotOwnedByUsers` should be removed and it should not be possible to withdraw ETH this way.
> 
> For tokens everything should work correctly as it is now. But let me think about this some more. :)

Done

**Comment by Brechtpd** [comment]
It seems like we don't need `NoDefaultFunc` any more:
> Contracts that receive Ether directly (without a function call, i.e. using send or transfer) but do not define a fallback function throw an exception, sending back the Ether (this was different before Solidity v0.4.0). So if you want your contract to receive Ether, you have to implement a payable fallback function.

- [Solidity docs](https://solidity.readthedocs.io/en/v0.5.3/contracts.html#fallback-function)

**Comment by dong77** [comment]
> It seems like we don't need `NoDefaultFunc` any more:
> 
> > Contracts that receive Ether directly (without a function call, i.e. using send or transfer) but do not define a fallback function throw an exception, sending back the Ether (this was different before Solidity v0.4.0). So if you want your contract to receive Ether, you have to implement a payable fallback function.
> 
> * [Solidity docs](https://solidity.readthedocs.io/en/v0.5.3/contracts.html#fallback-function)

@Brechtpd can you create a separate PR for this?

**Comment by Brechtpd** [comment]
> @Brechtpd can you create a separate PR for this?

Sure. But it's actually not used anywhere else (except for some old imports that should have been removed) so if it's removed here there's not much to do any more. But if you merge this one as is I'll make a PR that removes it completely.

**Comment by dong77** [comment]
> > @Brechtpd can you create a separate PR for this?
> 
> Sure. But it's actually not used anywhere else (except for some old imports that should have been removed) so if it's removed here there's not much to do any more. But if you merge this one as is I'll make a PR that removes it completely.

I'll remove it.

**Pull request merged**